w3-aggregation.md

```diff
@@ -0,0 +1,359 @@
+
+```mermaid
+sequenceDiagram
+participant Storefront as 🌐<br/><br/>did:web:web3.storage
+    participant Authority as 🌐<br/><br/>did:web:spade.storage
+
+    Storefront->>Authority: invoke `aggregate/get`
+```
+
+## Capabilities
+
+This section describes the capabilities that form the w3 aggregation protocol, along with the details relevant for invoking capabilities with a service provider.
+
+In this document, we will be looking at `spade.storage` as an implementer of the `aggregate/*` and `offer/*` protocol.
+
+### `aggregate/offer`
+
+A Storefront principal can invoke a capabilty to submit an aggregate ready for offers.
+
+> `did:web:web3.storage` invokes capability from `did:web:spade.storage`
+
+```json
+{
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:spade.storage",
+  "att": [{
+    "with": "did:web:web3.storage",
+    "can": "aggregate/offer",
+    "nb": {
+      "offer": {
+        "link": "bafy...aggregate",
+        "size": 110101,
+        "src": ["https://w3s.link/ipfs/bafy...aggregate"]
+      }
+    }
+  }],
+  "prf": [],
+  "sig": "..."
+}
+```
+
+This capability is invoked to submit a request to a broker service when an aggregate is ready for a Filecoin deal. The `nb.offer.link` is the DAG-CBOR CID that refers to a "Ferry" aggregate. It encodes a dag-cbor block with an array of entries representing all the CAR files to include in the aggregate. Its format is:
+
+```json
+[
+  {
+    "link": "bag...",
+    "size": 110101,
+    "commP": "commP...",
+    "src": ["https://.../bag(...).car"]
+  } 
+]
+```
+
+A `src` with URLs to fetch the aggregate offer is provided. In addition, the total `size` of every CAR within it is provided, both for convenience and consistency with the dag-cbor blocks.
+
+A receipt will be generated to acknowledge the received request. This receipt MUST contain an [effect](https://github.com/ucan-wg/invocation/#7-effect) with a subsequent task (`.fx.join` field) that is run when submitted aggregate is processed and either succeeds (implying that aggregate was accepted and deals will be arranged) or fail (with `error` describing a problem with an aggregate)
+
+```json
+{
+  "ran": "bafy...invocation",
+  "out": {
+    "ok": {
+      "status": "queued"
+    }
+  },
+  "fx": {
+    "join": { "/": "bafy...dequeue" }
+  },
+  "meta": {},
+  "iss": "did:web:spade.storage",
+  "prf": []
+}
+```
+
+### `aggregate/get`
+
+A Storefront principal can invoke a capability to get state of an accepted aggregate.
+
+> `did:web:web3.storage` invokes capability from `did:web:spade.storage`
+
+```json
+{
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:spade.storage",
+  "att": [{
+    "with": "did:web:web3.storage",
+    "can": "aggregate/get",
+    "nb": {
+      "link": "bafy...aggregate",
+    }
+  }],
+  "prf": [],
+  "sig": "..."
+}
+```
+
+Once this invocation is executed, a receipt is generated with the aggregate information:
```

Comment:
I think we need to consider two things here:

1. How does cashing affects this ? Ideally invocations are idempotent meaning same instruction should get same result (even if state has changed in the meantime) and `nnc` supposed to be utilized to cache miss. If we do not want to deal with caching we should specify additional field like current time and request that it be set to e.g. latest drand or a wall clock, that way out dated responses are less surprising, if you request state for old state that has not been computed prior you get an error saying it can not be computed, if we have it cached than response is for that time frame.

1. We should specify what happens if you request deal info for the aggregate that is in `queued` state.

Reply:
That is reasonable, lets just add sentence to state that.